src/lib/client-file-uploader.js

```diff
@@ -0,0 +1,447 @@
+	} catch ( e ) {
+		throw `Invalid response from cloud service. ${ e }`;
+	}
+
+	const { Code, Message } = parsedResponse.Error || {};
+	throw `Unable to upload to cloud storage. ${ JSON.stringify( { Code, Message } ) }`;
+}
+
+export async function uploadUsingMultipart( {
+	app,
+	basename,
+	fileMeta,
+	fileName,
+	organization,
+}: UploadArguments ) {
+	console.log( 'Uploading to S3 using the Multipart API.' );
+
+	const presignedCreateMultipartUpload = await getSignedUploadRequestData( {
+		organizationId: organization.id,
+		appId: app.id,
+		basename,
+		action: 'CreateMultipartUpload',
+	} );
+
+	const multipartUploadResponse = await fetch(
+		presignedCreateMultipartUpload.url,
+		presignedCreateMultipartUpload.options
+	);
+	const multipartUploadResult = await multipartUploadResponse.text();
+
+	console.log( { multipartUploadResult } );
+
+	// TODO is any hardening needed here?
+	const parser = new XmlParser( {
+		explicitArray: false,
+		ignoreAttrs: true,
+	} );
+
+	const parsedResponse = await parser.parseStringPromise( multipartUploadResult );
+
+	if ( parsedResponse.Error ) {
+		const { Code, Message } = parsedResponse.Error;
+		throw `Unable to create cloud storage object. Error: ${ JSON.stringify( { Code, Message } ) }`;
+	}
+
+	if (
+		! parsedResponse &&
+		parsedResponse.InitiateMultipartUploadResult &&
+		parsedResponse.InitiateMultipartUploadResult.UploadId
+	) {
+		throw `Unable to get Upload ID from cloud storage. Error: ${ multipartUploadResult }`;
+	}
+
+	const uploadId = parsedResponse.InitiateMultipartUploadResult.UploadId;
+
+	console.log( { uploadId } );
+
+	const parts = getPartBoundaries( fileMeta.size );
+	const partsWithHash = await hashParts( fileName, parts );
+	const etagResults = await uploadParts( {
+		basename,
+		fileName,
+		parts: partsWithHash,
+		uploadId,
+	} );
+	console.log( { etagResults } );
+
+	console.log( 'Completing the upload...' );
+	return completeMultipartUpload( {
+		basename,
+		uploadId,
+		etagResults: [
+			etagResults[ 0 ], // Only one ETag object is required, don't waste bandwidth pushing them all
+		],
+	} );
+}
+
+export async function getSignedUploadRequestData( {
+	organizationId,
+	appId,
+	basename,
+	action,
+	etagResults,
+	uploadId = undefined,
+	partNumber = undefined,
+}: GetSignedUploadRequestDataArgs ): Promise<Object> {
+	const { apiFetch } = await API();
+	const response = await apiFetch( '/upload/signed-url', {
+		method: 'POST',
+		body: { action, appId, basename, etagResults, organizationId, partNumber, uploadId },
+	} );
+
+	if ( response.status !== 200 ) {
+		throw await response.text();
+	}
+
+	return response.json();
+}
+
+export type FileMeta = {
+	basename: string,
+	md5: string,
+	fileName: string,
+	size: number,
+};
+
+export async function checkFileAccess( fileName: string ): Promise<void> {
+	// Node 8 doesn't have fs.promises, so fall back to this
+	return new Promise( ( resolve, reject ) => {
```

Comment:
If this provides the same objective as `fs.promises` then all good. Otherwise, I think we can use `fs.promises` given code is transpiled by babel and Node 8 reached the end of life already.

Reply:
We're still running tests in node 8:

https://github.com/Automattic/vip/blob/bbfbb8f9693aa2f6ba5411bf9695478c9e56cb80/.circleci/config.yml#L9

It was failing on `fs.promises` & this works around it.  If we're no longer supporting node 8, we should take it out & bump our minimum node version (e.g. here https://github.com/Automattic/vip/blob/bbfbb8f9693aa2f6ba5411bf9695478c9e56cb80/package.json#L46)

Reply:
+1 for bumping to at least node 10 (probably 12).
But.... we should probably do that in a separate PR to ensure we don't introduce another variable of change.